sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -5,114 +5,220 @@
+    private final long prefetch;
+    private final AtomicInteger wip = new AtomicInteger();
+    private final Queue<SynchronousReceiveWork> workQueue = new ConcurrentLinkedQueue<>();
+    private final Queue<ServiceBusReceivedMessageContext> bufferMessages = new ConcurrentLinkedQueue<>();
+    private final SynchronousReceiveWork initialWork;
+    private final AtomicLong remaining = new AtomicLong();
 
     private volatile Subscription subscription;
 
-    SynchronousMessageSubscriber(SynchronousReceiveWork work) {
-        this.work = Objects.requireNonNull(work, "'work' cannot be null.");
+    private SynchronousReceiveWork currentWork;
+    private Disposable timeoutOperation;
+
+    SynchronousMessageSubscriber(long prefetch, SynchronousReceiveWork initialWork) {
+        this.prefetch = prefetch;
+        this.initialWork = initialWork;
     }
 
     /**
      * On an initial subscription, will take the first work item, and request that amount of work for it.
-     *
      * @param subscription Subscription for upstream.
      */
     @Override
     protected void hookOnSubscribe(Subscription subscription) {
-        if (this.subscription == null) {
-            this.subscription = subscription;
-        }
-
-        logger.info("[{}] Pending: {}, Scheduling receive timeout task '{}'.", work.getId(), work.getNumberOfEvents(),
-            work.getTimeout());
+        this.subscription = subscription;
 
-        subscription.request(work.getNumberOfEvents());
+        logger.verbose("[{}] onSubscribe Pending: {}, Scheduling receive timeout task '{}'.", initialWork.getId(),
+            initialWork.getNumberOfEvents(), initialWork.getTimeout());
 
-        timer.schedule(new ReceiveTimeoutTask(work.getId(), this::dispose), work.getTimeout().toMillis());
+        // This will trigger subscription.request(N) and queue up the work
+        queueWork(initialWork);
     }
 
+    /**
+     *
+     * @return next work to process from queue.
+     */
+    private SynchronousReceiveWork getNextWorkAndRequest() {
+        // Now see if there is more queued up work
+        SynchronousReceiveWork nextWork = workQueue.poll();
+
+        if (nextWork != null) {
+
+            logger.verbose("[{}] Picking up next receive request.", nextWork.getId());
+
+            // timer to complete the current in case of timeout trigger
+            timeoutOperation = getTimeoutOperation(nextWork.getTimeout());
+
+            requestCredits(nextWork.getNumberOfEvents());
+        }
+        return nextWork;
+    }
     /**
      * Publishes the event to the current {@link SynchronousReceiveWork}. If that work item is complete, will dispose of
      * the subscriber.
-     *
-     * @param value Event to publish.
+     * @param message Event to publish.
      */
     @Override
-    protected void hookOnNext(ServiceBusReceivedMessageContext value) {
-        work.next(value);
+    protected void hookOnNext(ServiceBusReceivedMessageContext message) {
```

Comment:
There's a lot of logic in this method which will make it hard to debug in the future... why wouldn't the following below work? And have any management of currentWork, or emitting the next work in a single place, inside the `drainQueue` method.

```java
    @Override
    protected void hookOnNext(ServiceBusReceivedMessageContext message) {
        bufferMessages.add(message);
        drain();
    }
```